File: DBMS_Lab_Project/DataStructures/LFUCache.cpp
#include "LFUCache.h"
#include <sstream>
#include "../BinaryUtils.h"

LFUCache::LFUCache(int cap) : capacity(cap), size(0), minFreq(0) {}

void LFUCache::updateFrequency(const std::string& key) {
    int freq = keyToValFreq[key].second;
    auto it = keyToIterator[key];

    // 1. Удаляем ключ из списка старой частоты
    freqToKeys[freq].erase(it);

    // 2. Если список старой частоты опустел и это была minFreq,
    //    увеличиваем minFreq.
    if (freqToKeys[freq].empty() && freq == minFreq) {
        minFreq++;
    }

    // 3. Увеличиваем частоту ключа
    freq++;
    keyToValFreq[key].second = freq;

    // 4. Добавляем ключ в список новой частоты (в конец, т.к. он "recently used")
    freqToKeys[freq].push_back(key);
    keyToIterator[key] = --freqToKeys[freq].end();
}

void LFUCache::evict() {
    if (freqToKeys.empty()) return;

    // 1. Находим ключ для удаления (первый в списке minFreq, т.к. он LRU)
    std::string keyToEvict = freqToKeys[minFreq].front();
    freqToKeys[minFreq].pop_front();

    // 2. Удаляем его из всех карт
    keyToValFreq.erase(keyToEvict);
    keyToIterator.erase(keyToEvict);
}

void LFUCache::internal_set(const std::string& key, const std::string& value, int freq) {
    keyToValFreq[key] = {value, freq};
    freqToKeys[freq].push_back(key);
    keyToIterator[key] = --freqToKeys[freq].end();
    // При десериализации важно правильно установить minFreq
    if (minFreq == 0 || freq < minFreq) {
        minFreq = freq;
    }
    size++;
}

std::string LFUCache::get(const std::string& key) {
    if (keyToValFreq.find(key) == keyToValFreq.end()) {
        return "-1";
    }

    updateFrequency(key);
    return keyToValFreq[key].first;
}

void LFUCache::set(const std::string& key, const std::string& value) {
    if (capacity <= 0) return;

    if (keyToValFreq.find(key) != keyToValFreq.end()) {
        // Ключ уже существует, просто обновляем значение и частоту
        keyToValFreq[key].first = value;
        updateFrequency(key);
    } else {
        // Новый ключ
        if (size >= capacity) {
            // Кэш полон, нужно удалить самый редкий элемент
            evict();
            size--;
        }

        // Добавляем новый элемент с частотой 1
        keyToValFreq[key] = {value, 1};
        freqToKeys[1].push_back(key);
        keyToIterator[key] = --freqToKeys[1].end();
        minFreq = 1; // Сбрасываем minFreq на 1 для нового элемента
        size++;
    }
}

std::string LFUCache::serialize() const {
    std::stringstream ss;
    writeSize(ss, capacity);
    writeSize(ss, keyToValFreq.size());
    for (const auto& pair : keyToValFreq) {
        writeString(ss, pair.first);
        writeString(ss, pair.second.first);
        writeInt(ss, pair.second.second);
    }
    return ss.str();
}

void LFUCache::deserialize(const std::string& str) {
    keyToValFreq.clear();
    freqToKeys.clear();
    keyToIterator.clear();
    size = 0;
    minFreq = 0;
    capacity = 0;

    if(str.empty()) return;
    std::stringstream ss(str);
    size_t cap;
    readSize(ss, cap);
    capacity = static_cast<int>(cap);
    size_t count;
    readSize(ss, count);

    for(size_t i = 0; i < count; ++i) {
        std::string key, val;
        int freq;
        readString(ss, key);
        readString(ss, val);
        readInt(ss, freq);
        internal_set(key, val, freq);
    }
}

File: DBMS_Lab_Project_Go/internal/datastructures/lfu_cache.go
package datastructures

import (
	"bytes"
	"container/list"
)

type ValFreq struct {
	Value string
	Freq  int
}

type LFUCache struct {
	capacity      int
	size          int
	minFreq       int
	keyToValFreq  map[string]ValFreq
	freqToKeys    map[int]*list.List
	keyToIterator map[string]*list.Element
}

func NewLFUCache(cap int) *LFUCache {
	return &LFUCache{
		capacity:      cap,
		size:          0,
		minFreq:       0,
		keyToValFreq:  make(map[string]ValFreq),
		freqToKeys:    make(map[int]*list.List),
		keyToIterator: make(map[string]*list.Element),
	}
}

func (c *LFUCache) updateFrequency(key string) {
	valFreq := c.keyToValFreq[key]
	freq := valFreq.Freq
	c.keyToValFreq[key] = ValFreq{Value: valFreq.Value, Freq: freq + 1}

	// Remove from old freq list
	oldList := c.freqToKeys[freq]
	elem := c.keyToIterator[key]
	oldList.Remove(elem)
	if oldList.Len() == 0 {
		delete(c.freqToKeys, freq)
		if c.minFreq == freq {
			c.minFreq++
		}
	}

	// Add to new freq list
	newFreq := freq + 1
	if _, exists := c.freqToKeys[newFreq]; !exists {
		c.freqToKeys[newFreq] = list.New()
	}
	newList := c.freqToKeys[newFreq]
	newElem := newList.PushBack(key)
	c.keyToIterator[key] = newElem
}

func (c *LFUCache) evict() {
	if c.size == 0 {
		return
	}
	minList := c.freqToKeys[c.minFreq]
	evictElem := minList.Front() // LRU in this freq bucket
	evictKey := evictElem.Value.(string)
	minList.Remove(evictElem)
	if minList.Len() == 0 {
		delete(c.freqToKeys, c.minFreq)
	}

	delete(c.keyToValFreq, evictKey)
	delete(c.keyToIterator, evictKey)
	c.size--
}

func (c *LFUCache) Get(key string) string {
	if _, exists := c.keyToValFreq[key]; !exists {
		return "" // Or error
	}
	val := c.keyToValFreq[key].Value
	c.updateFrequency(key)
	return val
}

func (c *LFUCache) Set(key, value string) {
	if c.capacity == 0 {
		return
	}
	if _, exists := c.keyToValFreq[key]; exists {
		c.keyToValFreq[key] = ValFreq{Value: value, Freq: c.keyToValFreq[key].Freq}
		c.updateFrequency(key)
		return
	}

	if c.size >= c.capacity {
		c.evict()
	}

	c.keyToValFreq[key] = ValFreq{Value: value, Freq: 1}
	if _, exists := c.freqToKeys[1]; !exists {
		c.freqToKeys[1] = list.New()
	}
	c.freqToKeys[1].PushBack(key)
	c.keyToIterator[key] = c.freqToKeys[1].Back()
	c.minFreq = 1
	c.size++
}

func (c *LFUCache) Serialize() string {
	var buf bytes.Buffer
	WriteInt(&buf, c.capacity)
	WriteSize(&buf, len(c.keyToValFreq))
	for key, vf := range c.keyToValFreq {
		WriteString(&buf, key)
		WriteString(&buf, vf.Value)
		WriteInt(&buf, vf.Freq)
	}
	return buf.String()
}

func (c *LFUCache) Deserialize(str string) {
	c.size = 0
	c.minFreq = 0
	c.keyToValFreq = make(map[string]ValFreq)
	c.freqToKeys = make(map[int]*list.List)
	c.keyToIterator = make(map[string]*list.Element)

	if str == "" {
		return
	}
	buf := bytes.NewBufferString(str)
	c.capacity, _ = ReadInt(buf)
	count, _ := ReadSize(buf)
	for i := 0; i < count; i++ {
		key, _ := ReadString(buf)
		val, _ := ReadString(buf)
		freq, _ := ReadInt(buf)
		c.internalSet(key, val, freq)
	}
}

func (c *LFUCache) internalSet(key, value string, freq int) {
	c.keyToValFreq[key] = ValFreq{Value: value, Freq: freq}
	if _, exists := c.freqToKeys[freq]; !exists {
		c.freqToKeys[freq] = list.New()
	}
	c.freqToKeys[freq].PushBack(key)
	c.keyToIterator[key] = c.freqToKeys[freq].Back()
	c.size++
	// Recompute minFreq is hard here without scanning all freqs.
	// But usually deserialize is followed by usage.
	// Let's try to maintain minFreq.
	if c.minFreq == 0 || freq < c.minFreq {
		c.minFreq = freq
	}
}
